src/VisualStudio/Core/Def/Implementation/Remote/GlobalNotificationRemoteDeliveryService.cs

```diff
@@ -0,0 +1,146 @@
+        }
+
+        public void Dispose()
+        {
+            UnregisterGlobalOperationNotifications();
+        }
+
+        private void RegisterGlobalOperationNotifications()
+        {
+            var globalOperationService = _services.GetService<IGlobalOperationNotificationService>();
+            if (globalOperationService != null)
+            {
+                globalOperationService.Started += OnGlobalOperationStarted;
+                globalOperationService.Stopped += OnGlobalOperationStopped;
+            }
+        }
+
+        private void UnregisterGlobalOperationNotifications()
+        {
+            var globalOperationService = _services.GetService<IGlobalOperationNotificationService>();
+            if (globalOperationService != null)
+            {
+                globalOperationService.Started -= OnGlobalOperationStarted;
+                globalOperationService.Stopped -= OnGlobalOperationStopped;
+            }
+        }
+
+        private void OnGlobalOperationStarted(object sender, EventArgs e)
+        {
+            lock (_globalNotificationsGate)
+            {
+                _globalNotificationsTask = _globalNotificationsTask.SafeContinueWithFromAsync(
+                    continuation, _cancellationToken, TaskContinuationOptions.None, TaskScheduler.Default);
+            }
+
+            async Task<GlobalNotificationState> continuation(Task<GlobalNotificationState> previousTask)
+            {
+                // Can only transition from NotStarted->Started.  If we hear about
+                // anything else, do nothing.
+                if (previousTask.Result != GlobalNotificationState.NotStarted)
+                {
+                    return previousTask.Result;
+                }
+
+                var client = await RemoteHostClient.TryGetClientAsync(_services, _cancellationToken).ConfigureAwait(false);
+                if (client == null)
+                {
+                    return previousTask.Result;
+                }
+
+                await client.TryRunRemoteAsync(
+                    WellKnownServiceHubServices.CodeAnalysisService,
+                    nameof(IRemoteGlobalNotificationDeliveryService.OnGlobalOperationStarted),
+                    solution: null,
+                    Array.Empty<object>(),
+                    callbackTarget: null,
+                    _cancellationToken).ConfigureAwait(false);
+
+                return GlobalNotificationState.Started;
+            }
+        }
+
+        private void OnGlobalOperationStopped(object sender, GlobalOperationEventArgs e)
+        {
+            lock (_globalNotificationsGate)
+            {
+                _globalNotificationsTask = _globalNotificationsTask.SafeContinueWithFromAsync(
+                    continuation, _cancellationToken, TaskContinuationOptions.None, TaskScheduler.Default);
+            }
+
+            async Task<GlobalNotificationState> continuation(Task<GlobalNotificationState> previousTask)
```

Comment:
```suggestion
            return;

            async Task<GlobalNotificationState> continuation(Task<GlobalNotificationState> previousTask)
```

Reply:
IDE style is to have explicit exit above the local functions  that way the read knows tehre's no 'outer' code that executes beyond that point.  (same for the above method plx).

Reply:
I just moved the method. But I can add it. It looks odd though, maybe it would be better if the local function was above .

---
In reply to: [419027397](https://github.com/dotnet/roslyn/pull/43928#discussion_r419027397) [](ancestors = 419027397)

Reply:
> I just moved the method. But I can add it. It looks odd though, maybe it would be better if the local function was above .

In general, IDE style is:

1. local functions at the end.  This enables someone to immediately see the main-thread code of the function without having to go through the helpers first.
2. an explicit exit prior to the local functions.  This ensures that you can immediately see there's no additional code that runs.  And the local functions don't intersperse hte actual method code.
3. local functions are PascalCased at the IDE layer.  We tend think of them as 'scoped sibling helper **methods**' versus 'a local variable of a function type'.  
4. Optionally: sometimes we put `// local functions` above hte local functions.  This has helped clarify things in PRs where sometimes it looks like a sibling method, and it gets confusing (esp. with captures) when you can't immediatley tell that it's a local-function.

Reply:
Seems like it's better then to make it a method then because all of the above is just adding extra stuff to make it clearer.

Reply:
tbh, that's my general preference.  but i get the desire to occasionally have a helper method that is just scoped narrowly.  in practice though, if your class is well contained, just having these be helper methods is 100% fine (And how we survived for decades) :)

Reply:
Especially the need to add a comment means the style is bad.